api/src/main/java/org/pmiops/workbench/api/DiskController.java

```diff
@@ -0,0 +1,124 @@
+package org.pmiops.workbench.api;
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+import javax.inject.Provider;
+import org.pmiops.workbench.db.dao.WorkspaceDao;
+import org.pmiops.workbench.db.model.DbUser;
+import org.pmiops.workbench.db.model.DbWorkspace;
+import org.pmiops.workbench.exceptions.NotFoundException;
+import org.pmiops.workbench.leonardo.model.LeonardoListPersistentDiskResponse;
+import org.pmiops.workbench.model.Disk;
+import org.pmiops.workbench.model.EmptyResponse;
+import org.pmiops.workbench.notebooks.LeonardoNotebooksClient;
+import org.pmiops.workbench.utils.mappers.LeonardoMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class DiskController implements DiskApiDelegate {
+  private static final Logger log = Logger.getLogger(DiskController.class.getName());
+  private final LeonardoNotebooksClient leonardoNotebooksClient;
+  private final WorkspaceDao workspaceDao;
+  private final LeonardoMapper leonardoMapper;
+  private final Provider<DbUser> userProvider;
+  // timestamp for the date was created in ISO 8601 format
+  private static final String diskCreatedDateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'";
+
+  @Autowired
+  public DiskController(
+      LeonardoNotebooksClient leonardoNotebooksClient,
+      WorkspaceDao workspaceDao,
+      Provider<DbUser> userProvider,
+      LeonardoMapper leonardoMapper) {
+    this.leonardoNotebooksClient = leonardoNotebooksClient;
+    this.workspaceDao = workspaceDao;
+    this.userProvider = userProvider;
+    this.leonardoMapper = leonardoMapper;
+  }
+
+  @Override
+  public ResponseEntity<Disk> getDisk(String workspaceNamespace) {
+    String googleProject = lookupWorkspace(workspaceNamespace).getGoogleProject();
+    String pdNamePrefix = userProvider.get().getUserPDNamePrefix();
+    SimpleDateFormat dateFormatter = new SimpleDateFormat(diskCreatedDateFormat);
+    AtomicInteger activePDCnt = new AtomicInteger();
+    AtomicInteger failedPDCnt = new AtomicInteger();
+    Stream<LeonardoListPersistentDiskResponse> responseStream =
```

Comment:
Few comments/tip:
(1) General suggestion, for complicated logic like this, suggest to add more comment to explain the logic. 
(2) Not sure what date format Leo is using, can you just do `Instant.parse(r.getAuditInfo().getCreatedDate())`
(3) I am not clear if this API would return FAIL PD to UI. The API doc and method did not explain it. From line 70's r1, r2 sort/compare method, seems you would also return a failed one if that is the most recent one? What is the reason for that? Please add some comment to explain in the code.
(4) Log error on number of `FAILED > 0` might be confusing. It is possible that the failure happened last week, but we show that log every time the same user make a getDisk call.  Depends on your answer 3, if we might just want to log the fail if that is the most recent on, or we don't log that if we don't care FAIL status at all.
(5)[ Peek is mainly used for support debugging. ](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#peek-java.util.function.Consumer-) personal I usually don't use that. Depends on your answer 3, we might be able to find some alternative.  Depends on your answer 3, 4, I can try to give some suggestion.

Reply:
Thanks for (1)(2)(5) suggestions! Agree with them. For
(3) Now I want to return the FAIL status to the UI and let the user know if the most recent PD failed. Because if we just return a NotFound, the user might keep creating new ones. That would be a risky action.
(4) I think we need to deal with the failed PD even if it's failed a long time ago. Because we don't know if Leo would retry that one some time. We might need to ask what Leo's guarantee about that. But i think it's a safe action for us to warn the failed ones anyway.